### PR TITLE
[LG] fix duplicated tempplates issue

### PR DESCRIPTION
**Pull request by feich-ms**

### CODE REVIEW

File: libraries/Microsoft.Bot.Builder.LanguageGeneration/TemplateEngine.cs [hunk 1]
```diff
@@ -47,27 +47,40 @@ public TemplateEngine()
...
+                     if (!Path.IsPathRooted(importPath))
                      {
-                        // get full path for importPath relative to path which is doing the import.
-                        importPath = Path.GetFullPath(Path.Combine(Path.GetDirectoryName(filePath), id));
+                         // get full path for importPath relative to path which is doing the import.
+                         importPath = Path.GetFullPath(Path.Combine(Path.GetDirectoryName(filePath), id));
                      }
-                    return (File.ReadAllText(importPath), importPath);
+
+                     if (!filePathContainer.Contains(importPath))
```
Comment (Danieladu) [inline libraries/Microsoft.Bot.Builder.LanguageGeneration/TemplateEngine.cs]: We should keep `importResolver ` pure, and leave no other logic code in this delegate.  The only thing this delegate do is transfer from 'literal path' to 'file path'.

Comment (feich-ms) [reply libraries/Microsoft.Bot.Builder.LanguageGeneration/TemplateEngine.cs]: current importResolver also read file content and return file content and file path. You mean we change it to return file path only?

File: libraries/Microsoft.Bot.Builder.LanguageGeneration/TemplateEngine.cs [hunk 2]
```diff
@@ -220,13 +232,13 @@ private string GetOsPath(string ambigiousPath)
...
         /// </summary>
         /// <param name="content">Text content contains lg templates.</param>
         /// <param name="name">Text name.</param>
         /// <param name="importResolver">resolver to resolve LG import id to template text.</param>
         /// <returns>LGTemplate list of parsed lg content.</returns>
-        private IList<LGTemplate> ParseContent(string content, string name, ImportResolverDelegate importResolver)
+        private List<LGTemplate> ParseContent(string content, string name, ImportResolverDelegate importResolver)
         {
             var sources = new Dictionary<string, LGResource>();
             LoopLGText(content, name, sources, importResolver);
```
Comment (boydc2014) [inline libraries/Microsoft.Bot.Builder.LanguageGeneration/TemplateEngine.cs]: We should have a pure resource discovery function like
Dict<string, LGResource> =  DiscoverLGResoures(LGResource start, Resolver importResolver); // already de-dup
which make this more decoupled and simplify the whole design

File: libraries/Microsoft.Bot.Builder.LanguageGeneration/TemplateEngine.cs (hunk 1 above)
Comment (feich-ms) [reply libraries/Microsoft.Bot.Builder.LanguageGeneration/TemplateEngine.cs]: Removed fileContainer from import resolver

File: libraries/Microsoft.Bot.Builder.LanguageGeneration/TemplateEngine.cs (hunk 2 above)
Comment (feich-ms) [reply libraries/Microsoft.Bot.Builder.LanguageGeneration/TemplateEngine.cs]: Changed to DiscoverLGResources(LGResource start, ImportResolverDelegate importResolver, HashSet<string> parsedResoureceIds). Use the hashset to track visited files

File: libraries/Microsoft.Bot.Builder.LanguageGeneration/TemplateEngine.cs [hunk 3]
```diff
@@ -156,42 +160,85 @@ public string Evaluate(string inlineStr, object scope = null, IGetMethod methodB
...
+
+        /// <summary>
+        /// Resolve imported LG resources from a start resource.
+        /// All the imports will be visited and resolved to LGResouce list.
+        /// </summary>
+        /// <param name="start">The lg resource from which to start resolving imported resources.</param>
+        /// <param name="importResolver">resolver to resolve LG import id to template text.</param>
+        /// <param name="resourceIdsFound">Resource ids that have been vistied.</param>
+        /// <returns>LGTemplate list of parsed lg content.</returns>
+        private List<LGResource> ResolveImportResources(LGResource start, ImportResolverDelegate importResolver, HashSet<string> resourceIdsFound)
```
Comment (boydc2014) [inline libraries/Microsoft.Bot.Builder.LanguageGeneration/TemplateEngine.cs]: Can this just be HashSet<LGResource>?

File: libraries/Microsoft.Bot.Builder.LanguageGeneration/TemplateEngine.cs [hunk 4]
```diff
@@ -156,42 +160,85 @@ public string Evaluate(string inlineStr, object scope = null, IGetMethod methodB
...
+            var resourceIdsFound = new HashSet<string>();
+
             if (importResolver == null)
             {
                 // default to fileResolver...
                 importResolver = FileResolver;
             }
 
-            foreach (var id in ids)
+            return ResolveImportResources(start, importResolver, resourceIdsFound);
```
Comment (boydc2014) [inline libraries/Microsoft.Bot.Builder.LanguageGeneration/TemplateEngine.cs]: importResolver ?? FileResolver

File: libraries/Microsoft.Bot.Builder.LanguageGeneration/TemplateEngine.cs (hunk 3 above)
Comment (feich-ms) [reply libraries/Microsoft.Bot.Builder.LanguageGeneration/TemplateEngine.cs]: You mean the return value should be HashSet<LGResource>?

Comment (feich-ms) [reply libraries/Microsoft.Bot.Builder.LanguageGeneration/TemplateEngine.cs]: fixed

File: libraries/Microsoft.Bot.Builder.LanguageGeneration/TemplateEngine.cs (hunk 4 above)
Comment (feich-ms) [reply libraries/Microsoft.Bot.Builder.LanguageGeneration/TemplateEngine.cs]: fixed

File: libraries/Microsoft.Bot.Builder.LanguageGeneration/LGResource.cs [hunk 5]
```diff
@@ -43,5 +43,24 @@ public LGResource(IList<LGTemplate> templates, IList<LGImport> imports, string i
...
         public IList<LGImport> Imports { get; set; }
+
+        /// <summary>
+        /// Override the Equals function for LGResource comparison.
+        /// </summary>
+        /// <param name="obj">The LGResource object.</param>
+        /// <returns>True if the ids are same.</returns>
+        public override bool Equals(object obj)
+        {
+            return this.Id.Equals(((LGResource)obj).Id);
```
Comment (Danieladu) [inline libraries/Microsoft.Bot.Builder.LanguageGeneration/LGResource.cs]: if (LGResource)obj == null?

Comment (feich-ms) [reply libraries/Microsoft.Bot.Builder.LanguageGeneration/LGResource.cs]: fixed

**Pull request merged**